### PR TITLE
15150 bug deadlock found when trying to get lock on dataobject import 1

### LINKED ISSUE #15150: [Bug] - Deadlock found when trying to get lock on DataObject import
Hi, 

For your information, we are facing the exact same problem with the `object_meatada` table on one of our object. 

What happens is, we are trying to populate a fresh install of pimcore with a little bit more than 100 000 DataObjects through the DataImporter Bundle. These DataObjects have multiple relation fields. Some of these fields are filled through the DataImporter, but others are filled using an EventListener (using the `PreSaveEvent` of the DataImporter) because what we are trying to achieve is not possible natively with the DataImporter.

When we execute the import in parallel with multiple process (ie. 10), a lot of objects couldn't be created because of this error :
```
Next Doctrine\DBAL\Exception\DeadlockException: An exception occurred while executing 'INSERT INTO object_metadata_OBJECT (o_id, dest_id, fieldname, ownertype, ownername, index, position, type, column, data) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?) ON DUPLICATE KEY UPDATE o_id = ?, dest_id = ?, fieldname = ?, ownertype = ?, ownername = ?, index = ?, position = ?, type = ?, column = ?, data = ?;' with params [my params]:

SQLSTATE[40001]: Serialization failure: 1213 Deadlock found when trying to get lock; try restarting transaction in /var/www/html/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php:51
```

It failed even though it retried, according to the message.

The params (my enveloppes) are good since, to try to debug this, we imported all of our objects by executing the parallel process with 1 process (took 6 hours, but no errors). It is clearly linked to the parallelization of the import.

I also have to mention that we were able to import everything by commenting our event in the DataImporter, with multiple processes (again, 10). So the events in the DataImporter seems to be a part of the problem as well in our case. 

Lastly, the problem doesn't seem to appear when updating already created objects. It only affects the creation.

Should I create a new issue with this problem, or is it also part of this problem ?

_Originally posted by @SamyMP in https://github.com/pimcore/pimcore/issues/15064#issuecomment-1539650108_

### PR DESCRIPTION
**Pull request by robertSt7**
partially resolves #15150

### CODE REVIEW

File: bundles/CoreBundle/Migrations/Version20230616085142.php [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+        $metaDataTables = $db->fetchAllAssociative(
+            "SHOW FULL TABLES
+                       WHERE `Tables_in_{$db->getDatabase()}` LIKE 'object_metadata_%' AND Table_type = 'BASE TABLE'"
+        );
+        foreach ($metaDataTables as $table) {
+            $metaDataTable = $schema->getTable(current($table));
+
+            if ($metaDataTable->hasColumn(self::AUTO_ID)) {
+                $metaDataTable->dropPrimaryKey();
```
Comment (kingjia90) [inline bundles/CoreBundle/Migrations/Version20230616085142.php]: facing an issue when trying to `migrate --down`
To reproduce, migrate up, change an object to have a `adv.many to many` field and save it, trying to run migration down `bin/console d:m:e "Pimcore\Bundle\CoreBundle\Migrations\Version20230616085142" --down`
Only crashes on the freshly created table
image


probably needs
```suggestion
                $this->addSql('ALTER TABLE `' . $tableName . '` DROP FOREIGN KEY IF EXISTS `' . $fkName . '`');                
                $metaDataTable->dropPrimaryKey();
```
and readd it by addForeignKeyConstraint

**Review by kingjia90 — CHANGES_REQUESTED** [review]

File: bundles/CoreBundle/Migrations/Version20230616085142.php [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            $metaDataTable = $schema->getTable($tableName);
+
+            if (!$metaDataTable->hasColumn(self::AUTO_ID)) {
+                $metaDataTable->addColumn(self::AUTO_ID, 'integer', [
+                    'autoincrement' => true,
+                ]);
+
+                $fkName = AbstractDao::getForeignKeyName($tableName, self::ID_COLUMN);
+                $this->addSql('ALTER TABLE `' . $tableName . '` DROP FOREIGN KEY IF EXISTS `' . $fkName . '`');
+                $metaDataTable->dropPrimaryKey();
```
Comment (dvesh3) [inline bundles/CoreBundle/Migrations/Version20230616085142.php]: I would stick to one approach for the all changes. can we use `addSql()` consistently?

File: bundles/CoreBundle/Migrations/Version20230616085142.php (hunk 1 above)
Comment (kingjia90) [reply bundles/CoreBundle/Migrations/Version20230616085142.php]: Found a discrepancy by checking with PhpStorm SQL Generator: on the left, it's a migrated table and on the right is a created by `DAO::createOrUpdateTable()`

image
beside the `id` column order should be put at first position in migration and the difference in LIMIT `INT(11)`, there are `fk_object_metadata_NE__o_id` and `IDX_208F4545DB01246B` respectively different. This is likely the reason it crashes on migration down

File: bundles/CoreBundle/Migrations/Version20230616085142.php [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                       WHERE `Tables_in_{$db->getDatabase()}` LIKE 'object_metadata_%' AND Table_type = 'BASE TABLE'"
+            );
+        foreach ($metaDataTables as $table) {
+            $tableName = current($table);
+            $metaDataTable = $schema->getTable($tableName);
+
+            if (!$metaDataTable->hasColumn(self::AUTO_ID)) {
+                $this->addSql(
+                    'ALTER TABLE `' . $tableName . '` DROP PRIMARY KEY, ' .
+                    'ADD `' . self::AUTO_ID . '` int(11) UNSIGNED NOT NULL AUTO_INCREMENT PRIMARY KEY'
```
Comment (kingjia90) [inline bundles/CoreBundle/Migrations/Version20230616085142.php]:
```suggestion
                    'ADD `' . self::AUTO_ID . '` int(11) UNSIGNED NOT NULL AUTO_INCREMENT PRIMARY KEY FIRST'
```
to keep `id` as absolute first column

File: bundles/CoreBundle/Migrations/Version20230616085142.php [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+namespace Pimcore\Bundle\CoreBundle\Migrations;
+
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+use Pimcore\Db;
+
+final class Version20230616085142 extends AbstractMigration
+{
+    private const ID_COLUMN = 'o_id';
+    private const PK_COLUMNS = '`' . self::ID_COLUMN .
```
Comment (kingjia90) [inline bundles/CoreBundle/Migrations/Version20230616085142.php:26]: i'd keep `PK_COLUMNS` as an array and use a 
```
UNIQUE (`' . implode('`,', SELF:PK_COLUMNS) . '`)'
```
but not that important change, just looks easier to read and "mantain" the array (and sonarcloud line char length, breaking in multiple lines)

File: bundles/CoreBundle/Migrations/Version20230616085142.php [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            $metaDataTable = $schema->getTable($tableName);
+
+            if (!$metaDataTable->hasColumn(self::AUTO_ID)) {
+                $this->addSql(
+                    'ALTER TABLE `' . $tableName . '` DROP PRIMARY KEY, ' .
+                    'ADD `' . self::AUTO_ID . '` int(11) UNSIGNED NOT NULL AUTO_INCREMENT PRIMARY KEY'
+                );
+                $this->addSql(
+                    'ALTER TABLE `' . $tableName . '` ADD ' .
+                    'CONSTRAINT `' . self::UNIQUE_KEY_NAME . '` UNIQUE (' . self::PK_COLUMNS . ')'
```
Comment (kingjia90) [inline bundles/CoreBundle/Migrations/Version20230616085142.php]: mhhh, is that necessary? is it to trigger "rebuild" the foreign key or so? it seems doesn't change before/after

**Review by kingjia90 — CHANGES_REQUESTED** [review]
Overall LGTM and works now, just some minor inputs

File: bundles/CoreBundle/Migrations/Version20230616085142.php (hunk 5 above)
Comment (robertSt7) [reply bundles/CoreBundle/Migrations/Version20230616085142.php]: This is for adding the "old" primary key as unique constraint. Sorry I don't get what you exactly mean

File: bundles/CoreBundle/Migrations/Version20230616085142.php (hunk 4 above)
Comment (robertSt7) [reply bundles/CoreBundle/Migrations/Version20230616085142.php:26]: I also thought about this, but then it's harder to read in the sql string

Comment (dvesh3) [reply bundles/CoreBundle/Migrations/Version20230616085142.php:26]: yes, even the suggested approach doesn't work with quotes so let's keep it like this...

Comment (kingjia90) [reply bundles/CoreBundle/Migrations/Version20230616085142.php:26]: ok, fine with me, looks something that is not going to be changed so often anyways

File: bundles/CoreBundle/Migrations/Version20230616085142.php (hunk 5 above)
Comment (kingjia90) [reply bundles/CoreBundle/Migrations/Version20230616085142.php]: it seems it was already like that, even before this PR. After the migration, it doesn't change
https://github.com/pimcore/demo/blob/10.3/dump/data-0-bootstrap.sql#L1514

Comment (kingjia90) [reply bundles/CoreBundle/Migrations/Version20230616085142.php]: ahh, sorry, my mistake, it's totally fine

Comment (robertSt7) [reply bundles/CoreBundle/Migrations/Version20230616085142.php]: This is not the foreign key this is a unique key for `o_id` `dest_id`, `type`, `fieldname`, `column`, `ownertype`, `ownername`, `position`, `index`

**Review by kingjia90 — APPROVED** [review]

**Review by dvesh3 — APPROVED** [review]

**Comment by jeykodev** [comment]
@kingjia90 / @dvesh3 I just tried to Update from 11.0.0 to 11.0.7 and the migration is failing. 

`[error] Migration Pimcore\Bundle\CoreBundle\Migrations\Version20230616085142 failed during Execution. Error: "An exception occurred while executing a query: SQLSTATE[HY000]: General error: 1553 Cannot drop index 'PRIMARY': needed in a foreign key constraint"`

I can't keep up with the changes done to this migration, it's very confusing and I guess there are some side effects. Can you take a look?

**Comment by dvesh3** [comment]
@jeykodev thanks for pointing this out. We'll take a look into this issue.

**Comment by robertSt7** [comment]
@jeykodev We have released 11.0.8. Could you please try to re-migrate the Migration? Thanks

**Pull request merged**